Content.Server/Botany/Systems/MutationSystem.cs

```diff
@@ -0,0 +1,211 @@
+
+        // Hybrids have a high chance of being seedless. Balances very
+        // effective hybrid crossings.
+        if (a.Name == result.Name && random(0.7f))
+        {
+            result.Seedless = true;
+        }
+
+        return result;
+    }
+
+    // Mutate reference 'val' between 'min' and 'max' by pretending the value
+    // is representable by a thermometer code with 'bits' number of bits and
+    // randomly flipping some of them.
+    //
+    // 'totalbits' and 'mult' are used only to calculate the probability that
+    // one bit gets flipped.
+    private static void MutateFloat(ref float val, float min, float max, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        // Starting number of bits that are high, between 0 and n.
+        int n = (int)Math.Round((val - min) / (max - min) * bits);
+
+        // Probability that the bit flip increases n.
+        float p_increase = 1-(float)n/bits;
+        int np;
+        if (random(p_increase))
+        {
+            np = n + 1;
+        }
+        else
+        {
+            np = n - 1;
+        }
+
+        // Set value based on mutated thermometer code.
+        float nval = (float)Math.Min(Math.Max((float)np/bits * (max - min) + min, min), max);
+        val = nval;
+    }
+
+    private static void MutateInt(ref int n, int min, int max, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        // Probability that the bit flip increases n.
+        float p_increase = 1-(float)n/bits;
+        int np;
+        if (random(p_increase))
+        {
+            np = n + 1;
+        }
+        else
+        {
+            np = n - 1;
+        }
+
+        np = Math.Min(Math.Max(np, min), max);
+        n = np;
+    }
+
+    private static void MutateBool(ref bool val, bool polarity, int bits, int totalbits, float mult)
+    {
+        // Probability that a bit flip happens for this value.
+        float p = mult*bits/totalbits;
+        if (!random(p))
+        {
+            return;
+        }
+
+        val = polarity;
+    }
+
+    private static Color RandomColor(Color color, int bits, int totalbits, float mult)
+    {
+        float p = mult*bits/totalbits;
+        if (random(p))
+        {
+            var colors = new List<Color>{
+                Color.White,
+                Color.Red,
+                Color.Yellow,
+                Color.Green,
+                Color.Blue,
+                Color.Purple,
+                Color.Pink
+            };
+            return _robustRandom.Pick(colors);
+        }
+        return color;
+    }
```

Comment:
why not just truly randomize the color? was this tested and it just looked bad aesthetically (i know some color systems fare badly with pure randomization)

Reply:
These colors are interesting and distinguishable. With the small light radius, it can be hard to discern certain color glows (e.g. rolling grey would just show up as a dimmer white). As you point out, what I probably want here is randomizing only 'H' in the HSV color space, and converting back to RGB.